manifests/mongos.pp

```diff
@@ -1,26 +1,26 @@
 # This installs a Mongo Shard daemon. See README.md for more details.
 class mongodb::mongos (
-  $ensure           = $mongodb::params::mongos_ensure,
-  $config           = $mongodb::params::mongos_config,
-  $config_content   = undef,
-  $config_template  = undef,
-  $configdb         = $mongodb::params::mongos_configdb,
-  $config_data      = $mongodb::params::config_data,
-  $service_manage   = $mongodb::params::mongos_service_manage,
-  $service_provider = undef,
-  $service_name     = $mongodb::params::mongos_service_name,
-  $service_enable   = $mongodb::params::mongos_service_enable,
-  $service_ensure   = $mongodb::params::mongos_service_ensure,
-  $service_status   = $mongodb::params::mongos_service_status,
-  $package_ensure   = $mongodb::params::package_ensure_mongos,
-  $package_name     = $mongodb::params::mongos_package_name,
-  $unixsocketprefix = $mongodb::params::mongos_unixsocketprefix,
-  $pidfilepath      = $mongodb::params::mongos_pidfilepath,
-  $logpath          = $mongodb::params::mongos_logpath,
-  $fork             = $mongodb::params::mongos_fork,
-  $bind_ip          = undef,
-  $port             = undef,
-  $restart          = $mongodb::params::mongos_restart,
+  Variant[Boolean, String] $ensure                          = $mongodb::params::mongos_ensure,
+  Stdlib::Absolutepath $config                              = $mongodb::params::mongos_config,
+  $config_content                                           = undef,
+  $config_template                                          = undef,
+  $configdb                                                 = $mongodb::params::mongos_configdb,
+  Optional[Hash] $config_data                               = $mongodb::params::config_data,
+  Boolean $service_manage                                   = $mongodb::params::mongos_service_manage,
+  Optional[String] $service_provider                        = undef,
+  Optional[String] $service_name                            = $mongodb::params::mongos_service_name,
+  Boolean $service_enable                                   = $mongodb::params::mongos_service_enable,
+  Enum['stopped','running'] $service_ensure                 = $mongodb::params::mongos_service_ensure,
+  Optional[Enum['stopped','running']] $service_status       = $mongodb::params::mongos_service_status,
+  Variant[Boolean, String] $package_ensure                  = $mongodb::params::package_ensure_mongos,
+  String $package_name                                      = $mongodb::params::mongos_package_name,
+  Optional[Stdlib::Absolutepath] $unixsocketprefix          = $mongodb::params::mongos_unixsocketprefix,
+  Optional[Stdlib::Absolutepath] $pidfilepath               = $mongodb::params::mongos_pidfilepath,
+  Optional[Variant[Boolean, Stdlib::Absolutepath]] $logpath = $mongodb::params::mongos_logpath,
```

Comment:
Having a boolean here looks interesting

Reply:
I think it's needed  because there are cases where there are defaults in certain cases (from params), but users need to be able to suppress them? So in that case, passing `undef` will get the default, but `false` will work to suppress it. This is a quirk of Puppet that I don't particularly like, but other than adding an additional boolean parameter, I don't know if there's an easy fix.

Reply:
One option is ```Variant[Enum[''], Stdlib::Absolutepath]``` (or ```'unmanaged'```). These are times I miss the explicit ```null``` rather than ```undef```.